File: enrollment_system.py
"""
SAIT Enrollment Management System - Python Implementation
Part 3 of Assignment 2

This program implements a student enrollment management system that allows
administrative staff to add, modify, remove, and display student records.

This Python implementation replicates the functionality of the C++ version
developed in Part 2, demonstrating how the same algorithm can be implemented
in different programming languages with different typing systems and syntax.

Author: Altesse Imena
Date: October 23, 2025
"""

# Global variables for student data management
# Python uses dynamic typing - types are determined at runtime
student_records = []  # List to store all student records (dynamic array)
student_dict = {}     # Dictionary for quick lookup by student ID
next_student_id = 100001  # Auto-incrementing student ID counter


def display_menu():
    """
    Displays the main menu options
    Subprogram with no parameters (demonstrates modular design)
    Author: Altesse Imena
    """
    print("===== SAIT Enrollment Management System =====")
    print("1. Add Student Record")
    print("2. Modify Student Record")
    print("3. Remove Student Record")
    print("4. Display Student Record")
    print("5. Display All Student Records")
    print("6. Exit System")


def add_student():
    """
    Adds a new student record to the system
    Demonstrates dynamic typing and use of dictionaries
    Uses global variables to maintain state across function calls
    Author: Altesse Imena
    """
    global next_student_id  # Access global variable for modification
    
    # Create student dictionary to hold all information
    # Python dictionaries provide dynamic key-value storage
    student = {}
    
    # Assign unique student ID (auto-increment)
    student["id"] = next_student_id
    next_student_id += 1  # Increment for next student
    
    # Collect student information with input prompts
    student["first_name"] = input("Enter First Name: ")
    student["last_name"] = input("Enter Last Name: ")
    student["date_of_birth"] = input("Enter Date of Birth (YYYY-MM-DD): ")
    student["gender"] = input("Enter Gender: ")
    
    # Input validation for GPA (must be numeric and in valid range)
    while True:
        try:
            gpa = float(input("Enter GPA from Previous Institution: "))
            if 0.0 <= gpa <= 4.0:
                student["gpa"] = gpa
                break
            else:
                print("Invalid GPA. Please enter a value between 0.0 and 4.0.")
        except ValueError:
            print("Invalid input. Please enter a numeric value.")
    
    # Input validation for semester (must be positive integer)
    while True:
        try:
            semester = int(input("Enter Current Semester: "))
            if semester >= 1:
                student["semester"] = semester
                break
            else:
                print("Invalid semester. Please enter a positive number.")
        except ValueError:
            print("Invalid input. Please enter an integer.")
    
    student["program"] = input("Enter Program: ")
    
    # Input validation for number of courses
    while True:
        try:
            num_courses = int(input("Enter Number of Courses: "))
            if num_courses >= 0:
                student["num_courses"] = num_courses
                break
            else:
                print("Invalid number. Please enter a non-negative number.")
        except ValueError:
            print("Invalid input. Please enter an integer.")
    
    # Store student record in list
    student_records.append(student)
    
    # Store reference to student in dictionary for quick lookup
    # Python dictionaries store references to objects
    student_dict[student["id"]] = student
    
    print(f"Student record added successfully. Assigned Student ID: {student['id']}")


def modify_student():
    """
    Modifies an existing student record
    Demonstrates use of references and conditional statements
    Python passes mutable objects (like dictionaries) by reference
    Author: Altesse Imena
    """
    # Input validation for student ID
    try:
        search_id = int(input("Enter Student ID to modify: "))
    except ValueError:
        print("Invalid input.")
        return
    
    # Check if student ID exists in the dictionary
    if search_id in student_dict:
        # Get reference to student dictionary
        student = student_dict[search_id]
        
        print(f"Record found for ID {student['id']}: {student['first_name']} {student['last_name']}")
        
        # Display modification menu
        print("Select field to modify:")
        print("1. First Name")
        print("2. Last Name")
        print("3. Date of Birth")
        print("4. Gender")
        print("5. GPA")
        print("6. Semester")
        print("7. Program")
        print("8. Number of Courses")
        
        try:
            choice = int(input("Enter choice: "))
        except ValueError:
            print("Invalid input.")
            return
        
        # Conditional statements to handle different field modifications
        # Demonstrates branching logic
        if choice == 1:
            student["first_name"] = input("Enter new First Name: ")
        elif choice == 2:
            student["last_name"] = input("Enter new Last Name: ")
        elif choice == 3:
            student["date_of_birth"] = input("Enter new Date of Birth: ")
        elif choice == 4:
            student["gender"] = input("Enter new Gender: ")
        elif choice == 5:
            # Input validation for GPA
            while True:
                try:
                    gpa = float(input("Enter new GPA: "))
                    if 0.0 <= gpa <= 4.0:
                        student["gpa"] = gpa
                        break
                    else:
                        print("Invalid GPA. Please enter a value between 0.0 and 4.0.")
                except ValueError:
                    print("Invalid input. Please enter a numeric value.")
        elif choice == 6:
            # Input validation for semester
            while True:
                try:
                    semester = int(input("Enter new Semester: "))
                    if semester >= 1:
                        student["semester"] = semester
                        break
                    else:
                        print("Invalid semester. Please enter a positive number.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")
        elif choice == 7:
            student["program"] = input("Enter new Program: ")
        elif choice == 8:
            # Input validation for number of courses
            while True:
                try:
                    num_courses = int(input("Enter new Number of Courses: "))
                    if num_courses >= 0:
                        student["num_courses"] = num_courses
                        break
                    else:
                        print("Invalid number. Please enter a non-negative number.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")
        else:
            print("Invalid choice.")
            return
        
        print("Record updated successfully.")
    else:
        print("No record found with that Student ID.")


def remove_student():
    """
    Removes a student record from the system
    Demonstrates list and dictionary manipulation
    Author: Altesse Imena
    """
    # Input validation for student ID
    try:
        search_id = int(input("Enter Student ID of the student to remove: "))
    except ValueError:
        print("Invalid input.")
        return
    
    # Check if student ID exists
    if search_id in student_dict:
        # Get reference to student
        student = student_dict[search_id]
        
        # Remove from list using remove method
        student_records.remove(student)
        
        # Remove from dictionary
        del student_dict[search_id]
        
        print(f"Student record with ID {search_id} removed successfully.")
    else:
        print("No student found with that Student ID.")


def display_student():
    """
    Displays a single student record
    Demonstrates formatted output and conditional statements
    Author: Altesse Imena
    """
    # Input validation for student ID
    try:
        search_id = int(input("Enter Student ID to display: "))
    except ValueError:
        print("Invalid input.")
        return
    
    # Check if student exists in dictionary
    if search_id in student_dict:
        student = student_dict[search_id]
        
        # Display formatted student information
        print("----- Student Information -----")
        print(f"Student ID: {student['id']}")
        print(f"Name: {student['first_name']} {student['last_name']}")
        print(f"DOB: {student['date_of_birth']}")
        print(f"Gender: {student['gender']}")
        print(f"GPA: {student['gpa']:.2f}")  # Format to 2 decimal places
        print(f"Semester: {student['semester']}")
        print(f"Program: {student['program']}")
        print(f"Courses: {student['num_courses']}")
        print("-------------------------------")
    else:
        print("No student found with that Student ID.")


def display_all_students():
    """
    Displays all student records in the system
    Demonstrates use of loops (for-each) and conditional statements
    Author: Altesse Imena
    """
    # Check if there are any records
    if len(student_records) == 0:
        print("No student records found.")
        return
    
    print("----- Student Records -----")
    
    # Iterate through all students using for-each loop
    # Demonstrates iteration over a list
    for student in student_records:
        print(f"Student ID: {student['id']}")
        print(f"Name: {student['first_name']} {student['last_name']}")
        print(f"DOB: {student['date_of_birth']}")
        print(f"Gender: {student['gender']}")
        print(f"GPA: {student['gpa']:.2f}")
        print(f"Semester: {student['semester']}")
        print(f"Program: {student['program']}")
        print(f"Courses: {student['num_courses']}")
        print("----------------------------")


def main():
    """
    Main program entry point
    Implements the main menu loop for the enrollment system
    Author: Altesse Imena
    """
    running = True  # Control variable for main loop
    
    # Main program loop - continues until user chooses to exit
    while running:
        display_menu()
        
        # Input validation for menu choice
        try:
            choice = int(input("Enter your choice: "))
        except ValueError:
            print("Invalid input. Please enter a number.")
            print()
            continue
        
        # Conditional statements for menu navigation (demonstrates branching)
        # Python doesn't have switch statements (until 3.10 match-case)
        # Using if-elif-else chain instead
        if choice == 1:
            add_student()  # Call subprogram to add student
        elif choice == 2:
            modify_student()  # Call subprogram to modify student
        elif choice == 3:
            remove_student()  # Call subprogram to remove student
        elif choice == 4:
            display_student()  # Call subprogram to display single student
        elif choice == 5:
            display_all_students()  # Call subprogram to display all students
        elif choice == 6:
            print("Exiting system... Goodbye!")
            running = False  # Exit the loop
        else:
            print("Invalid option. Please try again.")
        
        print()  # Add spacing between operations


# Entry point for the program
# This ensures the main function only runs when the script is executed directly
if __name__ == "__main__":
    main()

File: enrollment_system.cpp
/*
 * SAIT Enrollment Management System - C++ Implementation
 * Part 2 of Assignment 2
 * 
 * This program implements a student enrollment management system that allows
 * administrative staff to add, modify, remove, and display student records.
 * 
 * Author: Altesse Imena
 * Date: October 23, 2025
 */

#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <iomanip>
#include <limits>

using namespace std;

// Student structure to hold all student information
// Demonstrates use of various data types (string, int, double)
struct Student {
    int id;                    // Unique student ID (static binding - type known at compile time)
    string first_name;         // Student's first name
    string last_name;          // Student's last name
    string date_of_birth;      // Date of birth in YYYY-MM-DD format
    string gender;             // Student's gender
    double gpa;                // GPA from previous institution (double for precision)
    int semester;              // Current semester number
    string program;            // Academic program name
    int num_courses;           // Number of courses enrolled
};

// Global variables for student data management
// Demonstrates static data binding (types known at compile time)
vector<Student> student_records;           // Dynamic array (list) to store all student records
map<int, Student*> student_dict;           // Dictionary (map) for quick lookup by ID using references
int next_student_id = 100001;              // Auto-incrementing student ID counter

// Function prototypes (forward declarations)
void display_menu();
void add_student();
void modify_student();
void remove_student();
void display_student();
void display_all_students();
void clear_input_buffer();

/*
 * Main program entry point
 * Implements the main menu loop for the enrollment system
 * Author: Altesse Imena
 */
int main() {
    int choice;  // Local variable with block scope
    bool running = true;  // Control variable for main loop
    
    // Main program loop - continues until user chooses to exit
    while (running) {
        display_menu();
        cout << "Enter your choice: ";
        
        // Input validation to handle non-integer input
        if (!(cin >> choice)) {
            cout << "Invalid input. Please enter a number." << endl;
            clear_input_buffer();
            continue;
        }
        
        clear_input_buffer();  // Clear remaining newline character
        
        // Switch statement for menu navigation (demonstrates branching)
        switch (choice) {
            case 1:
                add_student();  // Call subprogram to add student
                break;
            case 2:
                modify_student();  // Call subprogram to modify student
                break;
            case 3:
                remove_student();  // Call subprogram to remove student
                break;
            case 4:
                display_student();  // Call subprogram to display single student
                break;
            case 5:
                display_all_students();  // Call subprogram to display all students
                break;
            case 6:
                cout << "Exiting system... Goodbye!" << endl;
                running = false;  // Exit the loop
                break;
            default:
                cout << "Invalid option. Please try again." << endl;
        }
        
        cout << endl;  // Add spacing between operations
    }
    
    return 0;
}

/*
 * Displays the main menu options
 * Subprogram with no parameters (demonstrates modular design)
 * Author: Altesse Imena
 */
void display_menu() {
    cout << "===== SAIT Enrollment Management System =====" << endl;
    cout << "1. Add Student Record" << endl;
    cout << "2. Modify Student Record" << endl;
    cout << "3. Remove Student Record" << endl;
    cout << "4. Display Student Record" << endl;
    cout << "5. Display All Student Records" << endl;
    cout << "6. Exit System" << endl;
}

/*
 * Adds a new student record to the system
 * Demonstrates pass by value for primitive types and use of references
 * Author: Altesse Imena
 */
void add_student() {
    Student student;  // Local variable - demonstrates block scoping
    
    // Assign unique student ID (auto-increment)
    student.id = next_student_id;
    next_student_id++;  // Increment for next student
    
    // Collect student information with input prompts
    cout << "Enter First Name: ";
    getline(cin, student.first_name);
    
    cout << "Enter Last Name: ";
    getline(cin, student.last_name);
    
    cout << "Enter Date of Birth (YYYY-MM-DD): ";
    getline(cin, student.date_of_birth);
    
    cout << "Enter Gender: ";
    getline(cin, student.gender);
    
    // Input validation for GPA (must be numeric)
    cout << "Enter GPA from Previous Institution: ";
    while (!(cin >> student.gpa) || student.gpa < 0.0 || student.gpa > 4.0) {
        cout << "Invalid GPA. Please enter a value between 0.0 and 4.0: ";
        clear_input_buffer();
    }
    clear_input_buffer();
    
    // Input validation for semester (must be positive integer)
    cout << "Enter Current Semester: ";
    while (!(cin >> student.semester) || student.semester < 1) {
        cout << "Invalid semester. Please enter a positive number: ";
        clear_input_buffer();
    }
    clear_input_buffer();
    
    cout << "Enter Program: ";
    getline(cin, student.program);
    
    // Input validation for number of courses
    cout << "Enter Number of Courses: ";
    while (!(cin >> student.num_courses) || student.num_courses < 0) {
        cout << "Invalid number. Please enter a non-negative number: ";
        clear_input_buffer();
    }
    clear_input_buffer();
    
    // Store student record in vector (list)
    student_records.push_back(student);
    
    // Store reference to student in map (dictionary) for quick lookup
    // Demonstrates use of referencing - storing pointer to the last element
    student_dict[student.id] = &student_records.back();
    
    cout << "Student record added successfully. Assigned Student ID: " << student.id << endl;
}

/*
 * Modifies an existing student record
 * Demonstrates use of references and conditional statements
 * Author: Altesse Imena
 */
void modify_student() {
    int search_id;
    
    cout << "Enter Student ID to modify: ";
    if (!(cin >> search_id)) {
        cout << "Invalid input." << endl;
        clear_input_buffer();
        return;
    }
    clear_input_buffer();
    
    // Check if student ID exists in the dictionary (demonstrates referencing)
    if (student_dict.find(search_id) != student_dict.end()) {
        Student* student = student_dict[search_id];  // Get reference to student
        
        cout << "Record found for ID " << student->id << ": " 
             << student->first_name << " " << student->last_name << endl;
        
        // Display modification menu
        cout << "Select field to modify:" << endl;
        cout << "1. First Name" << endl;
        cout << "2. Last Name" << endl;
        cout << "3. Date of Birth" << endl;
        cout << "4. Gender" << endl;
        cout << "5. GPA" << endl;
        cout << "6. Semester" << endl;
        cout << "7. Program" << endl;
        cout << "8. Number of Courses" << endl;
        
        int choice;
        cout << "Enter choice: ";
        if (!(cin >> choice)) {
            cout << "Invalid input." << endl;
            clear_input_buffer();
            return;
        }
        clear_input_buffer();
        
        // Switch statement to handle different field modifications
        // Demonstrates branching and conditional logic
        switch (choice) {
            case 1:
                cout << "Enter new First Name: ";
                getline(cin, student->first_name);
                break;
            case 2:
                cout << "Enter new Last Name: ";
                getline(cin, student->last_name);
                break;
            case 3:
                cout << "Enter new Date of Birth: ";
                getline(cin, student->date_of_birth);
                break;
            case 4:
                cout << "Enter new Gender: ";
                getline(cin, student->gender);
                break;
            case 5:
                cout << "Enter new GPA: ";
                while (!(cin >> student->gpa) || student->gpa < 0.0 || student->gpa > 4.0) {
                    cout << "Invalid GPA. Please enter a value between 0.0 and 4.0: ";
                    clear_input_buffer();
                }
                clear_input_buffer();
                break;
            case 6:
                cout << "Enter new Semester: ";
                while (!(cin >> student->semester) || student->semester < 1) {
                    cout << "Invalid semester. Please enter a positive number: ";
                    clear_input_buffer();
                }
                clear_input_buffer();
                break;
            case 7:
                cout << "Enter new Program: ";
                getline(cin, student->program);
                break;
            case 8:
                cout << "Enter new Number of Courses: ";
                while (!(cin >> student->num_courses) || student->num_courses < 0) {
                    cout << "Invalid number. Please enter a non-negative number: ";
                    clear_input_buffer();
                }
                clear_input_buffer();
                break;
            default:
                cout << "Invalid choice." << endl;
                return;
        }
        
        cout << "Record updated successfully." << endl;
    } else {
        cout << "No record found with that Student ID." << endl;
    }
}

/*
 * Removes a student record from the system
 * Demonstrates use of iterators and container manipulation
 * Author: Altesse Imena
 */
void remove_student() {
    int search_id;
    
    cout << "Enter Student ID of the student to remove: ";
    if (!(cin >> search_id)) {
        cout << "Invalid input." << endl;
        clear_input_buffer();
        return;
    }
    clear_input_buffer();
    
    // Check if student ID exists
    if (student_dict.find(search_id) != student_dict.end()) {
        // Remove from vector using iterator
        // Demonstrates use of loops and conditional logic
        for (auto it = student_records.begin(); it != student_records.end(); ++it) {
            if (it->id == search_id) {
                student_records.erase(it);
                break;
            }
        }
        
        // Remove from dictionary
        student_dict.erase(search_id);
        
        // Update all pointers in the dictionary after vector modification
        // This is necessary because vector elements may have been moved in memory
        student_dict.clear();
        for (size_t i = 0; i < student_records.size(); ++i) {
            student_dict[student_records[i].id] = &student_records[i];
        }
        
        cout << "Student record with ID " << search_id << " removed successfully." << endl;
    } else {
        cout << "No student found with that Student ID." << endl;
    }
}

/*
 * Displays a single student record
 * Demonstrates formatted output and conditional statements
 * Author: Altesse Imena
 */
void display_student() {
    int search_id;
    
    cout << "Enter Student ID to display: ";
    if (!(cin >> search_id)) {
        cout << "Invalid input." << endl;
        clear_input_buffer();
        return;
    }
    clear_input_buffer();
    
    // Check if student exists in dictionary
    if (student_dict.find(search_id) != student_dict.end()) {
        Student* student = student_dict[search_id];
        
        // Display formatted student information
        cout << "----- Student Information -----" << endl;
        cout << "Student ID: " << student->id << endl;
        cout << "Name: " << student->first_name << " " << student->last_name << endl;
        cout << "DOB: " << student->date_of_birth << endl;
        cout << "Gender: " << student->gender << endl;
        cout << fixed << setprecision(2);  // Format GPA to 2 decimal places
        cout << "GPA: " << student->gpa << endl;
        cout << "Semester: " << student->semester << endl;
        cout << "Program: " << student->program << endl;
        cout << "Courses: " << student->num_courses << endl;
        cout << "-------------------------------" << endl;
    } else {
        cout << "No student found with that Student ID." << endl;
    }
}

/*
 * Displays all student records in the system
 * Demonstrates use of loops (for-each) and conditional statements
 * Author: Altesse Imena
 */
void display_all_students() {
    // Check if there are any records
    if (student_records.empty()) {
        cout << "No student records found." << endl;
        return;
    }
    
    cout << "----- Student Records -----" << endl;
    
    // Iterate through all students using range-based for loop
    // Demonstrates use of loops and pass by reference (const reference for efficiency)
    for (const Student& student : student_records) {
        cout << "Student ID: " << student.id << endl;
        cout << "Name: " << student.first_name << " " << student.last_name << endl;
        cout << "DOB: " << student.date_of_birth << endl;
        cout << "Gender: " << student.gender << endl;
        cout << fixed << setprecision(2);
        cout << "GPA: " << student.gpa << endl;
        cout << "Semester: " << student.semester << endl;
        cout << "Program: " << student.program << endl;
        cout << "Courses: " << student.num_courses << endl;
        cout << "----------------------------" << endl;
    }
}

/*
 * Utility function to clear input buffer
 * Prevents input errors from affecting subsequent input operations
 * Author: Altesse Imena
 */
void clear_input_buffer() {
    cin.clear();  // Clear error flags
    cin.ignore(numeric_limits<streamsize>::max(), '\n');  // Discard remaining input
}
